contracts/token/ERC20Channels.sol

```diff
@@ -0,0 +1,303 @@
+   */
+  function getChannelInfo(
+    address sender,
+    address receiver,
+    uint8 nonce
+  ) external view returns (bytes32, uint256, uint256, uint256) {
+    bytes32 channelId = getChannelId(sender, receiver, nonce);
+    require(channels[channelId].nonce > 0);
+
+    return (
+      channelId,
+      channels[channelId].deposit,
+      closingRequests[channelId].closeTime,
+      closingRequests[channelId].closingBalance
+    );
+  }
+
+  /*
+   *  Public functions
+   */
+
+  /**
+   * @dev Get the signer of a balance hash signed
+   * @param receiver address, the receiver to hash
+   * @param nonce uint8, the nonce number of the channel
+   * @param balance uint256, the balance to hash
+   * @param msgSigned bytes, the balance hash signed
+   */
+  function getSignerOfBalanceHash(
+    address receiver,
+    uint8 nonce,
+    uint256 balance,
+    bytes msgSigned
+  ) public view returns (address) {
+    bytes32 msgHash = generateBalanceHash(receiver, nonce, balance);
+    // Derive address from signature
+    address signer = msgHash.recover(msgSigned);
+    return signer;
+  }
+
+  /**
+   * @dev Generate a hash balance for an address
+   * @param receiver address, the receiver to hash
+   * @param nonce uint8, the nonce number of the channel
+   * @param balance uint256, the balance to hash
+   */
+  function generateBalanceHash(
+    address receiver,
+    uint8 nonce,
+    uint256 balance
+  ) public view returns (bytes32) {
+    return keccak256(receiver, nonce, balance, address(this));
+  }
+
+  /**
+   * @dev Generate a keccak256 hash
+   * @param message bytes, the mesage to hash
+   */
+  function generateKeccak256(bytes message) public pure returns(bytes32) {
+    return keccak256(message);
+  }
+
+  /**
+   * @dev Generate a channel id
+   * @param sender address, the sender in the channel
+   * @param receiver address, the receiver in the channel
+   * @param nonce uint8, the nonce number of the channel
+   */
+  function getChannelId(
+    address sender,
+    address receiver,
+    uint8 nonce
+  ) public pure returns (bytes32 data) {
+    return keccak256(sender, receiver, nonce);
+  }
+
+  /*
+   *  Internal functions
+   */
+
+  /**
+   * @dev Close a channel
+   * @param sender address, the sender in the channel
+   * @param receiver address, the receiver in the channel
+   * @param nonce uint8, the nonce number of the channel
+   * @param receiverBalance uint256, the final balance of the receiver
+   */
+  function close(
+    address sender,
+    address receiver,
+    uint8 nonce,
+    uint256 receiverBalance
+  ) internal {
+    bytes32 channelId = getChannelId(sender, receiver, nonce);
+    Channel memory channel = channels[channelId];
+
+    require(channel.nonce > 0);
+    require(receiverBalance <= channel.deposit);
+
+    // Remove closed channel structures
+    // channel.nonce will become 0
+    // Change state before transfer call
+    delete channels[channelId];
```

Comment:
Is it safe to reset `channel.nonce` to 0 after closing the contract? It would allow another channel to be created with the same nonce, for which the previous set of offline signed messages are still valid. Though only a `sender` can create a channel, and he should not do this as it is only in his own prejudice (as it allows the receiver to use the previous signed messages to extract the balance), should we add a safeguard to prevent the sender from that doing that mistake?